File: ts/src/day1/day1_b.ts
import * as fs from "fs";
import readline from "readline";

async function getDepths(): Promise<number[]> {
	const rl = readline.createInterface({
		input: fs.createReadStream("./input"),
	});

	const depths: number[] = [];

	for await (const line of rl) {
		const cur = Number.parseInt(line);
		depths.push(cur);
	}
	return new Promise<number[]>((resolve) => resolve(depths));
}

function countNumOfIncDepths(depths: number[]): number {
	let count = 0;

	let prevWindowSum = 0;
	let curWindowSum = 0;
	let windowSize = 0;

	for (let i = 0; i < depths.length; i++) {
		if (windowSize < 3) {
			windowSize++;
			curWindowSum += depths[i];
		} else {
			prevWindowSum = curWindowSum;
			curWindowSum = curWindowSum + depths[i] - depths[i - 3];

			if (curWindowSum > prevWindowSum) count++;
		}
	}

	return count;
}

(function main() {
	getDepths().then((depths) => console.log(countNumOfIncDepths(depths)));
})();

File: golang/day1/day1_b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func getDepths() []int {
	var depths []int

	file, _ := os.Open("./input")
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		cur, _ := strconv.Atoi(scanner.Text())

		depths = append(depths, cur)
	}

	return depths
}

func countNumOfIncDepths() int {
	count := 0
	depths := getDepths()

	windowSize := 0
	curWindowsSum := 0
	prevWindowsSum := 0

	for i := 0; i < len(depths); i++ {
		if windowSize < 3 {
			windowSize++
			curWindowsSum += depths[i]
		} else {
			prevWindowsSum = curWindowsSum
			curWindowsSum = curWindowsSum + depths[i] - depths[i-3]

			if curWindowsSum > prevWindowsSum {
				count++
			}
		}
	}

	return count
}

func main() {
	fmt.Println(countNumOfIncDepths())
}
